Play silence to keep audio device awake

mltony opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review". 
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Fixes #14386
### Summary of the issue:
Play silence in order to keep audio device open
### Description of user facing changes
By default most users won't notice anything as default volume is 0 (playing silence).
I added two config options in the advanced panel: silence duration (default = 30 seconds) and white noise volume (default is set to 0). Playing white noise would be helpfulin order to debug any user issues related to audio dropout.
### Description of development approach
Most of heavylifting has  been done by @jcsteh.
In `wasapi.cpp` there has been added `class SilencePlayer` that spawns a new thread, that just plays silence/white noise when requested. Requests are coming from python code in `nvwave.py` from `WasapiWavePlayer.feed()` function.

### Testing strategy:
Tested with white noise that it is being played as expected. Tested that config changes reflect immediately.
Couldn't test with silence as apparently my bluetooth headphones are not suffering from dropout problem in the first place.
### Known issues with pull request:
N/A
### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

jcsteh committed:
[WIP] WASAPI: Play silence to keep the audio device alive.

mltony committed:
White noise

cary-rowen commented:
I'll test this after you fix the CI build fail

lukaszgo1 commented on nvdaHelper/local/wasapi.cpp:
```diff
@@ -187,7 +190,6 @@ class WasapiPlayer {
 	HRESULT pause();
 	HRESULT resume();
 	HRESULT setChannelVolume(unsigned int channel, float level);
-
```
Why this change?

lukaszgo1 commented on user_docs/en/userGuide.t2t:
```diff
@@ -2594,6 +2594,16 @@ This option allows you to specify if NVDA will play an error sound in case an er
+
+==== White noise volume slider ====[whiteNoiseVolume]
+
+When you set white noise volume to any value above zero, then instead of silence NVDA would playh white noise in order to keep audio device open. This feature is useful for debugging audio issues.
```
```suggestion
When you set white noise volume to any value above zero, then instead of silence NVDA would play white noise in order to keep audio device open. This feature is useful for debugging audio issues.
```

lukaszgo1 commented on source/core.py:
Please re-initialize `nvwave` when user restores config to saved / defaults.

mltony replied:
No need to reinitialize since it reloads current settings for duration and volume automatically.

CyrilleB79 commented:
To me, these options seem too specific / technical for most of the users, including some users impacted by bluetooth device audio cut-off. Not every user knows what a white noise is, and from a user point of view, playing silence or with noise is not very meaningful.

> Playing white noise would be helpfulin order to debug any user issues related to audio dropout.

First of all, if white noise is only for debug purpose, it should not be in audio settings IMO, maybe just in advanced settings, or better, in an add-on.
If instead white noise is required by some device or sound cards, I think that we should just enforce white noise for everybody. In any case this white noise should not be audible, so it should not make a difference with full silence.

Personally, I would see just one option "Duration of audio inactivity (in seconds) before closing the audio stream". Or even simpler: a checkbox "Keep the audio device/stream always open". Then the in the User Guide, give information to the user to help setting this value, i.e.:
* too low = audio may be cut-off more often with a Bluetooth device
* too high = the battery discharges faster.

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Łukasz Golonka <lukasz.golonka@mailbox.org>

mltony committed:
Addressing comments

mltony commented:
@cary-rowen,
Feel free to test this now. installer NVDA test ids failing for everyone.

@CyrilleB79 

> these options seem too specific / technical for most of the users
> it should not be in audio settings 

I already put both settings in advanced panel. Most users should be happy with default values.

Some users complained to Jamie that this patch didn't fix their audio issue. We can either try to guess what happened to those users. Or alternatively we can ask them to switch to white noise and this will give us more information, because:
1. If white noise is not audible - something is really broken.
2. If white noise is audible but from another audio output device, then there is a problem with synchronization.
3. If white noise is audible from correct audio output, then something is wrong with auido 
If you take away white noise option, then there won't be a good way to debug those issues. 

> • too low = audio may be cut-off more often with a Bluetooth device
> • too high = the battery discharges faster.

Updated doc.

CyrilleB79 commented:
Sorry, @mltony, I had overlooked your PR and have not checked where the options were located.

cary-rowen commented:
Hi @mltony 
I tested this and it worked with my bluetooth headphones, and in the past I needed to use the bluetooth audio add-on.

Thanks

beqabeqa473 commented:
these all should be configurable to not play silence or whitenoise,
because most of the wireless headsets have a possibility to turn them
all upon inactivity time.

On 1/28/24, Rowen ***@***.***> wrote:
> Hi @mltony
> I tested this and it worked with my bluetooth headphones, and in the past I
> needed to use the bluetooth audio add-on.
>
> Thanks
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/16099#issuecomment-1913450600
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

mltony commented:
@beqabeqa473

Silence duration and white noise volume are both configurable.

> most of the wireless headsets have a possibility to turn them all upon inactivity time.

Not sure I understand this comment. Most Bluetooth headphones enter standby mode soon after each utterance  which causes speech dropout. As far as I know there is typically no way to configure standby mode on vast majority of Bluetooth headphones.

beqabeqa473 commented:
No, i ment 2.4G wireless headset, they are beeing turned off after
some period of inactivity which is configurable, it is 10 minutes for
me, and i don't this to go away, that is why i said that this options
should be configurable to turn them all together.

On 1/28/24, mltony ***@***.***> wrote:
> @beqabeqa473
>
> Silence duration and white noise volume are both configurable.
>
>> most of the wireless headsets have a possibility to turn them all upon
>> inactivity time.
>
> Not sure I understand this comment. Most Bluetooth headphones enter standby
> mode soon after each utterance  which causes speech dropout. As far as I
> know there is typically no way to configure standby mode on vast majority of
> Bluetooth headphones.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/16099#issuecomment-1913464386
> You are receiving this because you were mentioned.
>
> Message ID: ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

seanbudd commented:
I believe the white noise parameter should not be settable from the GUI, as its for dev debugging only, but keeping it has a hidden setting in nvda.ini may be appropriate

mltony committed:
Addressing comment by @seanbudd

mltony commented:
@seanbudd, I just dropped white noise volume from settings panel.

cary-rowen commented:
If the white noise volume is only used for development and debugging, it should be recorded in an appropriate doc. Otherwise, even developers will not know how to use it because it is too hidden.

mltony committed:
Update doc

cary-rowen commented:
If I remember correctly, reporting progress also has a hidden interval that can be configured.
This information should be documented, such as in the "Advanced Topics" section of the user guide.

CyrilleB79 commented on source/gui/settingsDialogs.py:
```diff
@@ -3295,6 +3295,23 @@ def __init__(self, parent):
 
 		self.Layout()
 
+		silenceDurationLabelText = _(
```
There is an "audio" grouping in the advanced settings panel. Could you add this parameter there rather than at the end? And update the documentation the same way?

CyrilleB79 commented on source/gui/settingsDialogs.py:
```diff
@@ -3295,6 +3295,23 @@ def __init__(self, parent):
 
+		silenceDurationLabelText = _(
+			# Translators: The label for a setting in advanced panel to change silence duration
+			"Duration in seconds of silence or white noise played to keep audio device open (or 0 to disable)"
```
1. I would totally drop "white noise" from the GUI. I'd recommend to add it in the developer's guide with other hidden config parameters (maybe in a separate PR though).
2. It's quite obvious that setting to 0 will add no silence at all. You can add "0 to disable" in the User Guide, but IMO, the GUI should not be cluttered with this detail.

```suggestion
			"Duration in seconds of silence to keep audio device open"
```

mltony committed:
Update source/gui/settingsDialogs.py

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Addressing comments by @CyrilleB79

mltony committed:
Merge branch 'master' into silence2

CyrilleB79 commented on source/core.py:
```diff
@@ -946,6 +945,7 @@ def _doPostNvdaStartupAction():
 			)
 		except:
 			pass
+	_terminate(nvwave)
```
Why was it moved? Could you update the initial description to give the reason if it is needed?

CyrilleB79 commented on user_docs/en/changes.t2t:
```diff
@@ -8,9 +8,9 @@ What's New in NVDA
 == New Features ==
 - In Windows 11, NVDA will announce alerts from voice typing and suggested actions including the top suggestion when copying data such as phone numbers to the clipboard (Windows 11 2022 Update and later). (#16009, @josephsl)
 - Added support for the BrailleEdgeS2 braille device. (#16033)
+- Playing silence after every utterance in order to keep audio device open (#14386, @jcsteh, @mltony)
```
Wouldn't you put something more user oriented?

E.g.:
```suggestion
- NVDA will play silence after every utterance in order to prevent the start of next speech to be clipped with some audio devices such as Bluetooth headphones. (#14386, @jcsteh, @mltony)
```

But this may probably be reworded still better.

CyrilleB79 commented on user_docs/en/userGuide.t2t:
```diff
@@ -2584,6 +2584,15 @@ Disabling WASAPI will disable the following options:
 
+==== Silence duration ====[silenceDuration]
+
+This edit box specifies how long NVDA plays silence after every utterance. We play silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
```
we -> NVDA

```suggestion
This edit box specifies how long NVDA plays silence after every utterance. NVDA plays silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
```

CyrilleB79 commented:
And also a question:
In [Jaws What's new](https://support.freedomscientific.com/Downloads/JAWS/JAWSWhatsNew?version=2020) presenting of the similar feature, I can read:
> This option is also not available if the Audio Ducking feature is enabled which lowers the audio volume of programs while JAWS speaks.

Have you checked this here? Is there a side effect?

XLTechie commented:
I have a question for @mltony or others:

I do not need this at all on my internal speakers. They have their 
own power saving feature, but it doesn't interfere with speech.

However, I do have bluetooth devices with very aggressive power saving, where 
this feature is needed. (and works well, thanks @mltony!)

It is both potentially costly to my laptop battery, and unnecessary, to use this 
feature when using the on-board speakers.

My question therefore is: could or should this be made configurable on a 
per-device basis?
I'm not sure if that would be tricky in the context of the Windows sound mapper, 
but not even most of my bluetooth headphones need this feature. Thus it would be 
nice to deploy it to only those which do.

That said, even if this is thought valid, I wouldn't suggest delaying the PR to 
achieve it. I'm also not sure how the UI for this might look.

CyrilleB79 commented:
I wonder if the effect of keeping the audio stream open on the laptop battery is really significant. @jcsteh have you done or heard from such benchmark in the past?

If yes, the issue raised by @XLTechie is completely valid. Though, implementing a per device configuration should be tracked in a separate issue and a subsequent PR IMO.
In the meantime, it will still be possible to assign a bluetooth profile that you may want to assign manually.

And if it is confirmed that the effect is significant on the laptop battery of some configurations, I'd say that the configuration of this feature should be moved from Advanced panel to Audio panel, unlike what I wrote before.

XLTechie commented:
@CyrilleB79 I'm not really setup to accurately test the battery of the laptop 
question, but I can probably more effectively figure out a way to test the 
battery of various bluetooth devices which don't need, but are subjected to, 
this feature.

Unfortunately I'm still recovering from some health problems I've had since 
Christmas, so I doubt I can do it in the next several days.

mltony committed:
Update source/gui/settingsDialogs.py

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Update user_docs/en/changes.t2t

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony commented:
@CyrilleB79 
Just checked ducking - appears to work fine with sound split.
@XLTechie
That's a valid argument. I would think that the impact on your laptop battery life should be minimal with default 30 second silence value. But feel free to test that.

mltony committed:
Addressing comments by @CyrilleB79

mltony committed:
lint

seanbudd commented on nvdaHelper/local/wasapi.cpp:
```diff
@@ -624,9 +634,128 @@ HRESULT WasapiPlayer::setChannelVolume(unsigned int channel, float level) {
+	for (UINT32 i = 0; i < n; i++) {
+		whiteNoiseData[i] = (INT16)dist(generator);
+	}
+}
+HRESULT SilencePlayer::init() {
```
```suggestion
}

HRESULT SilencePlayer::init() {
```

seanbudd commented on source/config/configSpec.py:
```diff
@@ -54,6 +54,8 @@
 	WASAPI = featureFlag(optionsEnum="BoolFlag", behaviorOfDefault="enabled")
 	soundVolumeFollowsVoice = boolean(default=false)
 	soundVolume = integer(default=100, min=0, max=100)
+	silenceTimeSeconds = integer(default=30,min=0,max=3600)
```
```suggestion
	silenceTimeSeconds = integer(default=30, min=0, max=3600)
```

CyrilleB79 replied:
If you accept to replace "play silence" by "Keep audio device awake" in the User Guide and the GUI, you may also change it here for consistency:

```suggestion
	keepAudioAwakeTimeSeconds = integer(default=30,min=0,max=3600)
```

seanbudd commented on user_docs/en/changes.t2t:
```diff
@@ -8,9 +8,9 @@ What's New in NVDA
 == New Features ==
 - In Windows 11, NVDA will announce alerts from voice typing and suggested actions including the top suggestion when copying data such as phone numbers to the clipboard (Windows 11 2022 Update and later). (#16009, @josephsl)
 - Added support for the BrailleEdgeS2 braille device. (#16033)
+- NVDA will play silence after every utterance in order to prevent the start of next speech to be clipped with some audio devices such as Bluetooth headphones. (#14386, @jcsteh, @mltony)
```
```suggestion
- NVDA will play silence after every utterance in order to prevent the start of the next speech being clipped with some audio devices such as Bluetooth headphones. (#14386, @jcsteh, @mltony)
```

seanbudd commented:
What are your thoughts on rephrasing this in the UX to "keep audio device awake" rather than "play silence"?

I think this is a bit more user friendly and outlines in the intended purpose better.
Thoughts @Qchristensen ?

Qchristensen commented:
> ...rephrasing this in the UX to "keep audio device awake" rather than "play silence"...

I'm in favour of this change.  As you note, "keep audio device awake" is descriptive of WHAT will happen if I enable that option.  "Play silence" is the HOW NVDA will do that, which, while some users may be interested in that, isn't as clear.  (It may be in the future there is some better way of keeping the audio device awake which doesn't involve explicitly sending silent sounds to it, in which case the first option will still be correct, but the second may not be.

CyrilleB79 commented on user_docs/en/userGuide.t2t:
```diff
@@ -2584,6 +2584,15 @@ Disabling WASAPI will disable the following options:
 
+==== Silence duration ====[SilenceDuration]
+
+This edit box specifies how long NVDA plays silence after every utterance. NVDA plays silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
```
Given the previous suggestion, we should replace "playing audio" by "keeping audio awake".

CyrilleB79 reviewed (commented):
I fully agree with replacing "Play silence" by "Keep audio device awake" in the GUI and in the User Guide for the reason indicated by @Qchristensen

mltony committed:
Update nvdaHelper/local/wasapi.cpp

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update source/config/configSpec.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update source/core.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user_docs/en/changes.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Addressing comments

mltony commented:
Renamed to "keep audio device awake" in UX as suggested. In user guide I renamed the section as well, but I left explanation that this is achieved by playing silence. Hope that's ok.

mltony committed:
Merge branch 'master' into silence2

mltony committed:
lint

Adriani90 commented:
@mltony and @jcsteh thanks very much for this nice work guys.

in the user guide, I think it makes sense to replace the word utterance with "speech sequence". This is easier to understand especially for people who need easier language. @mltony what do you think?
Also I would mention that this does not affect only bluetooth devices positively, but also systems with lower performance, and virtual machines such as virtual desktops in Citrix environments.

mltony commented:
@Adriani90, updated utterance -> speech sequence.
Re other use cases, what evidence do we have that silence positively affects systems with lower performance? My wife is running NVDA on a modest 8-year old laptop and she doesn't experience any problems with NVDA audio. I guess there must be users who are running NVDA on even older computers. But then I'd think just being low performance wouldn't necessarily imply bad audio quality - NVDA in general doesn't consume  much resources at all. Maybe there is some old technology like a specific brand of motherboards that suffers from this problem? I don't know whether we should go into that rabbithole of investigating which old equipment causes dropout in NVDA speech, but I would prefer to have some evidence (like user testimonial) that silence indeed helps on their old equipment.
Same thing for citrix. I previously ran NVDA in VMWare Player and  VirtualBox and I recall audio quality wasn't that perfect, but it didn't improve from Bluetooth Audio add-on. And the quality problems wer different: every few seconds there was some audible glitch in sound instead of beginnings of each utterance dropping, which would have been a sign of audio device being closed. My guess was that there was high latency between host and guest that caused occasional dropout of some small chunk of audio buffer. And in my understanding this own't be fixed by this PR.

mltony committed:
docs

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -2605,14 +2605,17 @@ Disabling WASAPI will disable the following options:
+==== Duration of keeping audio device awake ====[KeepAudioAwakeDuration]
 
-This edit box specifies how long NVDA plays silence after every utterance. NVDA plays silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
+This edit box specifies how long NVDA keeps audio device awake after every speech sequence.
```
```suggestion
This edit box specifies how long NVDA keeps the audio device awake after every speech sequence.
```

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -2605,14 +2605,17 @@ Disabling WASAPI will disable the following options:
 
-This edit box specifies how long NVDA plays silence after every utterance. NVDA plays silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
+This edit box specifies how long NVDA keeps audio device awake after every speech sequence.
+NVDA does that by playing silence to keep audio device open and avoid certain glitches like dropped parts of speech sequences.
```
```suggestion
NVDA does this by playing silence to keep the audio device open and avoid certain glitches like dropped parts of speech sequences.
```

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -2605,14 +2605,17 @@ Disabling WASAPI will disable the following options:
 
-You can specify silence duration to be 0 in order to disable this feature. This slider is only availabel when wasapi mode is enabled.
+You can specify silence duration to be 0 in order to disable this feature.
+This slider is only available when WASAPI mode is enabled.
```
```suggestion
This slider is only available when [WASAPI mode #WASAPI] is enabled.
```

seanbudd committed:
Apply suggestions from code review

Adriani90 commented:
At least in Citrix I am using NVDA at work and without the bluetooth addon the speech is cutting and crackling alot. I am not using any wireless headset, but a wired headfone. The bluetooth addon definitely has a positive effect on this. Without it the citrix virtual desktop is literarly unusable for me.  Von meinem iPhone gesendetAm 07.02.2024 um 23:38 schrieb mltony ***@***.***>:﻿
@Adriani90, updated utterance -> speech sequence.
Re other use cases, what evidence do we have that silence positively affects systems with lower performance? My wife is running NVDA on a modest 8-year old laptop and she doesn't experience any problems with NVDA audio. I guess there must be users who are running NVDA on even older computers. But then I'd think just being low performance wouldn't necessarily imply bad audio quality - NVDA in general doesn't consume  much resources at all. Maybe there is some old technology like a specific brand of motherboards that suffers from this problem? I don't know whether we should go into that rabbithole of investigating which old equipment causes dropout in NVDA speech, but I would prefer to have some evidence (like user testimonial) that silence indeed helps on their old equipment.
Same thing for citrix. I previously ran NVDA in VMWare Player and  VirtualBox and I recall audio quality wasn't that perfect, but it didn't improve from Bluetooth Audio add-on. And the quality problems wer different: every few seconds there was some audible glitch in sound instead of beginnings of each utterance dropping, which would have been a sign of audio device being closed. My guess was that there was high latency between host and guest that caused occasional dropout of some small chunk of audio buffer. And in my understanding this own't be fixed by this PR.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

mltony committed:
docs

mltony commented:
@Adriani90 , fair enough, I added a line about that in the user guide. Now I also vaguely remember someone on NVDA mailing list was saying their laptop's built-in audio was suffering from audio cutoffs and Bluetooth add-on indeed helped, so I mentioned that as well.

Qchristensen reviewed (approved):
All reads well, good work.

jcsteh commented:
@mltony Thanks very much for driving this to completion. :)

XLTechie commented:
I'm curious why it is thought that "speech sequence" makes sense to end users? 
Wouldn't something like "each spoken phrase", or describing it as "after speech 
stops" be more comprehensible?

Even the original, "utterance", comes across as a more user friendly word, 
although perhaps some other languages have issues with it as @Adriani90 implied.

I would rather see us describe it with as few technical terms as possible. 
"Speech sequence" strikes me as a term of art, which is not well defined 
outside our developer community.
